src/hotspot/share/memory/metaspace/binList.hpp

```diff
@@ -156,13 +154,13 @@ class BinListImpl {
       Block* b = _blocks[index];
       const size_t real_word_size = word_size_for_index(index);
       assert(b != nullptr, "Sanity");
-      assert(b->_word_size >= word_size &&
-             b->_word_size == real_word_size,
-             "bad block size in list[%d] (" BLOCK_FORMAT ")", index, BLOCK_FORMAT_ARGS(b));
       _blocks[index] = b->_next;
       _counter.sub(real_word_size);
       *p_real_word_size = real_word_size;
-      return (MetaWord*)b;
+      MetaWord* p = (MetaWord*)b;
+      assert(real_word_size == 1 || p[real_word_size - 1] == 0,
```

Comment:
Isn't the first case also handled by the second??

Reply:
This probably would deserve a comment, but I recently got obsessed with patch sizes :/

This line checks the trailing canary of this block that is written in BinList::add_block(). But if the blocks are of size 1, there is no space for the canary (in BinList::add_block(), the canary is overwritten by the 1-word-sized Block Header). Therefore, if the block is size 1, we cannot check the canary.